In [179]:
import pandas as pd

In [180]:
pd.set_option("display.max_columns", None)

# Data capec multilabel dataset

In [181]:
df_capec = pd.read_csv('Dataset/data_capec_multilabel.csv')

C:\Users\anhnmt2\AppData\Local\Temp\ipykernel_14988\1247042814.py:1: DtypeWarning: Columns (9,11,12,13,14,15,16,18,19) have mixed types. Specify dtype option on import or set low_memory=False.
  df_capec = pd.read_csv('Dataset/data_capec_multilabel.csv')


In [182]:
df_capec.head()

timestamp      src_ip  src_port      dst_ip  dst_port  \
0  17/Jul/2020:12:23:34 +0100  172.26.0.1   55894.0  172.26.0.4      80.0   
1  17/Jul/2020:12:23:41 +0100  172.26.0.1   55897.0  172.26.0.4      80.0   
2  17/Jul/2020:12:23:51 +0100  172.26.0.1   55901.0  172.26.0.4      80.0   
3  17/Jul/2020:12:23:51 +0100  172.26.0.1   55902.0  172.26.0.4      80.0   
4  17/Jul/2020:12:23:58 +0100  172.26.0.1   55903.0  172.26.0.4      80.0   

  request_http_method                               request_http_request  \
0                 GET                                                  /   
1                 GET  /blog/index.php/2020/04/04/voluptatum-reprehen...   
2                 GET                               /blog/xmlrpc.php?rsd   
3                 GET                                                  /   
4                 GET  /blog/index.php/2020/04/04/nihil-tenetur-et-ad...   

  request_http_protocol                                 request_user_agent  \
0              HTTP/1.1  Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_3...   
1              HTTP/1.1  Mozilla/5.0 (Macintosh; Intel Mac OS X 10.6; r...   
2              HTTP/1.1  Mozilla/4.0 (compatible; MSIE 6.0; Windows NT ...   
3              HTTP/1.1  Mozilla/5.0 (Linux; Android 4.2.2; SM-T110 Bui...   
4              HTTP/1.1  Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_2...   

  request_referer   request_host request_origin request_cookie  \
0             NaN  test-site.com            NaN            NaN   
1             NaN  test-site.com            NaN            NaN   
2             NaN  test-site.com            NaN            NaN   
3             NaN  test-site.com            NaN            NaN   
4             NaN  test-site.com            NaN            NaN   

  request_content_type request_accept request_accept_language  \
0                  NaN            */*                     NaN   
1                  NaN            */*                     NaN   
2                  NaN            */*                     NaN   
3                  NaN            */*                     NaN   
4                  NaN            */*                     NaN   

  request_accept_encoding  request_do_not_track request_connection  \
0           gzip, deflate                   NaN         keep-alive   
1           gzip, deflate                   NaN         keep-alive   
2           gzip, deflate                   NaN         keep-alive   
3           gzip, deflate                   NaN         keep-alive   
4           gzip, deflate                   NaN         keep-alive   

  request_body response_http_protocol  response_http_status_code  \
0          NaN               HTTP/1.1                      200.0   
1          NaN               HTTP/1.1                      200.0   
2          NaN               HTTP/1.1                      200.0   
3          NaN               HTTP/1.1                      200.0   
4          NaN               HTTP/1.1                      200.0   

  response_http_status_message  response_content_length  000 - Normal  \
0                           OK                  25174.0             1   
1                           OK                  20862.0             1   
2                           OK                    317.0             1   
3                           OK                  25174.0             1   
4                           OK                  20863.0             1   

   272 - Protocol Manipulation  242 - Code Injection  \
0                            0                     0   
1                            0                     0   
2                            0                     0   
3                            0                     0   
4                            0                     0   

   88 - OS Command Injection  126 - Path Traversal  66 - SQL Injection  \
0                          0                     0                   0   
1                          0                     0                   0   
2              

In [183]:
## Delete blank text in http request
df_capec = df_capec[df_capec['request_http_request'] != '/']

## Convert from multilabel to multiclass

In [184]:
df_capec.shape

(608806, 38)

In [185]:
def count_duplicate(a0,a1,a2,a3,a4,a5,a6,a7,a8,a9,a10,a11,a12,a13):
    return a0+a1+a2+a3+a4+a5+a6+a7+a8+a9+a10+a11+a12+a13
        
df_capec['count'] = df_capec.apply(lambda x: count_duplicate(x['000 - Normal'],x['272 - Protocol Manipulation'], x['242 - Code Injection'],x['88 - OS Command Injection'],x['126 - Path Traversal'],x['66 - SQL Injection'],x['16 - Dictionary-based Password Attack'],x['310 - Scanning for Vulnerable Software'],x['153 - Input Data Manipulation'],x['248 - Command Injection'],x['274 - HTTP Verb Tampering'],x['194 - Fake the Source of Data'],x['34 - HTTP Response Splitting'],x['33 - HTTP Request Smuggling']), axis = 1)

In [186]:
# Data with only multiclass

In [187]:
df_multiclass = df_capec[df_capec['count']==1]

In [188]:
df_multiclass.shape

(599755, 39)

In [189]:
df_multiclass.head()

timestamp      src_ip  src_port      dst_ip  dst_port  \
1  17/Jul/2020:12:23:41 +0100  172.26.0.1   55897.0  172.26.0.4      80.0   
2  17/Jul/2020:12:23:51 +0100  172.26.0.1   55901.0  172.26.0.4      80.0   
4  17/Jul/2020:12:23:58 +0100  172.26.0.1   55903.0  172.26.0.4      80.0   
5  17/Jul/2020:12:24:07 +0100  172.26.0.1   55910.0  172.26.0.4      80.0   
6  17/Jul/2020:12:24:15 +0100  172.26.0.1   55911.0  172.26.0.4      80.0   

  request_http_method                               request_http_request  \
1                 GET  /blog/index.php/2020/04/04/voluptatum-reprehen...   
2                 GET                               /blog/xmlrpc.php?rsd   
4                 GET  /blog/index.php/2020/04/04/nihil-tenetur-et-ad...   
5                 GET  /blog/index.php/2020/04/04/explicabo-qui-fuga-...   
6                 GET  /blog/index.php/2020/04/04/explicabo-qui-fuga-...   

  request_http_protocol                                 request_user_agent  \
1              HTTP/1.1  Mozilla/5.0 (Macintosh; Intel Mac OS X 10.6; r...   
2              HTTP/1.1  Mozilla/4.0 (compatible; MSIE 6.0; Windows NT ...   
4              HTTP/1.1  Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_2...   
5              HTTP/1.1  Mozilla/5.0 (Windows NT 10.0; Win64; x64) Appl...   
6              HTTP/1.1  Mozilla/5.0 (Windows NT 5.1) AppleWebKit/537.3...   

  request_referer   request_host request_origin request_cookie  \
1             NaN  test-site.com            NaN            NaN   
2             NaN  test-site.com            NaN            NaN   
4             NaN  test-site.com            NaN            NaN   
5             NaN  test-site.com            NaN            NaN   
6             NaN  test-site.com            NaN            NaN   

  request_content_type request_accept request_accept_language  \
1                  NaN            */*                     NaN   
2                  NaN            */*                     NaN   
4                  NaN            */*                     NaN   
5                  NaN            */*                     NaN   
6                  NaN            */*                     NaN   

  request_accept_encoding  request_do_not_track request_connection  \
1           gzip, deflate                   NaN         keep-alive   
2           gzip, deflate                   NaN         keep-alive   
4           gzip, deflate                   NaN         keep-alive   
5           gzip, deflate                   NaN         keep-alive   
6           gzip, deflate                   NaN         keep-alive   

  request_body response_http_protocol  response_http_status_code  \
1          NaN               HTTP/1.1                      200.0   
2          NaN               HTTP/1.1                      200.0   
4          NaN               HTTP/1.1                      200.0   
5          NaN               HTTP/1.1                      200.0   
6          NaN               HTTP/1.1                      200.0   

  response_http_status_message  response_content_length  000 - Normal  \
1                           OK                  20862.0             1   
2                           OK                    317.0             1   
4                           OK                  20863.0             1   
5                           OK                  21062.0             1   
6                           OK                  21061.0             1   

   272 - Protocol Manipulation  242 - Code Injection  \
1                            0                     0   
2                            0                     0   
4                            0                     0   
5                            0                     0   
6                            0                     0   

   88 - OS Command Injection  126 - Path Traversal  66 - SQL Injection  \
1                          0                     0                   0   
2                          0                     0                   0   
4              

In [190]:
# Data with multilabel

In [191]:
df_multilabel = df_capec[df_capec['count']>1]

In [192]:
df_multilabel.shape

(9051, 39)

In [193]:
df_multilabel.head()

timestamp      src_ip  src_port      dst_ip  dst_port  \
2940  17/Jul/2020:19:01:39 +0200  172.26.0.1   45500.0  172.26.0.4      80.0   
2961  17/Jul/2020:19:11:23 +0200  172.26.0.1   47482.0  172.26.0.4      80.0   
2969  17/Jul/2020:19:25:15 +0200  172.26.0.1   59862.0  172.26.0.4      80.0   
2974  17/Jul/2020:19:25:16 +0200  172.26.0.1   59862.0  172.26.0.4      80.0   
2977  17/Jul/2020:19:25:16 +0200  172.26.0.1   59862.0  172.26.0.4      80.0   

     request_http_method                               request_http_request  \
2940                POST  /vendor/phpunit/phpunit/src/Util/PHP/eval-stdi...   
2961                POST                               /cgi-bin/ViewLog.asp   
2969                 GET                                        /.svn/wc.db   
2974                 GET                                   /blog/.svn/wc.db   
2977                 GET              /blog/index.php/my-account/.svn/wc.db   

     request_http_protocol                                 request_user_agent  \
2940              HTTP/1.1  Mozilla/5.0 (Windows NT 10.0; Win64; x64) Appl...   
2961              HTTP/1.1                                 B4ckdoor-owned-you   
2969              HTTP/1.1  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
2974              HTTP/1.1  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
2977              HTTP/1.1  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   

     request_referer   request_host request_origin  \
2940             NaN  172.26.0.1:80            NaN   
2961             NaN      127.0.0.1            NaN   
2969             NaN  test-site.com            NaN   
2974             NaN  test-site.com            NaN   
2977             NaN  test-site.com            NaN   

                                         request_cookie  \
2940                                                NaN   
2961                                                NaN   
2969                                                NaN   
2974  wordpress_logged_in_1aefbe2f76edd740f8e362f39d...   
2977  wordpress_logged_in_1aefbe2f76edd740f8e362f39d...   

                   request_content_type request_accept  \
2940  application/x-www-form-urlencoded            NaN   
2961  application/x-www-form-urlencoded            */*   
2969                                NaN            NaN   
2974                                NaN            NaN   
2977                                NaN            NaN   

     request_accept_language request_accept_encoding  request_do_not_track  \
2940                     NaN                    gzip                   NaN   
2961                     NaN           gzip, deflate                   NaN   
2969                     NaN                     NaN                   NaN   
2974                     NaN                     NaN                   NaN   
2977                     NaN                     NaN                   NaN   

     request_connection                                       request_body  \
2940              close                                <?=md5("phpunit")?>   
2961         keep-alive   remote_submit_Flag=1&remote_syslog_Flag=1&Rem...   
2969                NaN                                                NaN   
2974                NaN                                                NaN   
2977                NaN                                                NaN   

     response_http_protocol  response_http_status_code  \
2940               HTTP/1.1                      404.0   
2961               HTTP/1.1                      404.0   
2969               HTTP/1.1                      404.0   
2974               HTTP/1.1                      404.0   
2977               HTTP/1.1                      404.0   

     response_http_status_message  response_content_length  000 - Normal  \
2940                    Not Found                    275.0             0   
2961                    Not Found                    271.0             0   
2969                    Not Found   

## 66 - SQL Injection

In [194]:
# Assign for test only
df_capec = df_multiclass

In [195]:
df_sqli = df_capec[df_capec['66 - SQL Injection'] == 1].loc[:, ['request_http_method','request_http_request', 'request_body', '66 - SQL Injection']].reset_index(drop=True)

In [196]:
df_sqli = df_capec[df_capec['66 - SQL Injection'] == 1]

In [197]:
df_sqli.head()

timestamp      src_ip  src_port      dst_ip  dst_port  \
3121  17/Jul/2020:19:25:48 +0200  172.26.0.1   59867.0  172.26.0.4      80.0   
4255  17/Jul/2020:19:33:20 +0200  172.26.0.1   59968.0  172.26.0.4      80.0   
7159  17/Jul/2020:20:01:20 +0200  172.26.0.1   63063.0  172.26.0.4      80.0   
7758  17/Jul/2020:21:27:43 +0200  172.26.0.1   64594.0  172.26.0.4      80.0   
8188  17/Jul/2020:21:29:12 +0200  172.26.0.1   64609.0  172.26.0.4      80.0   

     request_http_method                               request_http_request  \
3121                 GET  /blog/wp-admin/load-styles.php?c=0&dir=ltr&loa...   
4255                 GET  /blog/wp-admin/load-styles.php?c=0&dir=ltr&loa...   
7159                 GET  /blog/wp-admin/load-styles.php?c=0&dir=ltr&loa...   
7758                 GET  /blog/wp-admin/load-styles.php?c=0&dir=ltr&loa...   
8188                 GET  /blog/wp-admin/load-styles.php?c=0&dir=ltr&loa...   

     request_http_protocol                                 request_user_agent  \
3121              HTTP/1.1  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
4255              HTTP/1.1  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
7159              HTTP/1.1  Mozilla/5.0 (X11; Linux x86_64; rv:73.0) Gecko...   
7758              HTTP/1.1  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
8188              HTTP/1.1  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   

                                        request_referer   request_host  \
3121             http://test-site.com/blog/wp-login.php  test-site.com   
4255                                                NaN  test-site.com   
7159  https://test-site.com/blog/wp-login.php?logged...  test-site.com   
7758  http://test-site.com/blog/wp-login.php?action=...  test-site.com   
8188                                                NaN  test-site.com   

     request_origin                         request_cookie  \
3121            NaN  wordpress_test_cookie=WP+Cookie+check   
4255            NaN  wordpress_test_cookie=WP+Cookie+check   
7159            NaN  wordpress_test_cookie=WP+Cookie+check   
7758            NaN  wordpress_test_cookie=WP+Cookie+check   
8188            NaN  wordpress_test_cookie=WP+Cookie+check   

     request_content_type      request_accept request_accept_language  \
3121                  NaN                 NaN                     NaN   
4255                  NaN                 NaN                     NaN   
7159                  NaN  text/css,*/*;q=0.1          en-US,en;q=0.5   
7758                  NaN                 NaN                     NaN   
8188                  NaN                 NaN                     NaN   

     request_accept_encoding  request_do_not_track request_connection  \
3121                     NaN                   NaN                NaN   
4255                     NaN                   NaN                NaN   
7159                     NaN                   1.0         keep-alive   
7758                     NaN                   NaN                NaN   
8188                     NaN                   NaN                NaN   

     request_body response_http_protocol  response_http_status_code  \
3121          NaN               HTTP/1.1                      200.0   
4255          NaN               HTTP/1.1                      200.0   
7159          NaN               HTTP/1.1                      200.0   
7758          NaN               HTTP/1.1                      200.0   
8188          NaN               HTTP/1.1                      200.0   

     response_http_status_message  response_content_length  000 - Normal  \
3121                           OK                      NaN             0   
4255                           OK                      NaN             0   
7159                           OK                      NaN             0   
7758                           OK                      NaN             0   
8188                           OK                      NaN             0   



In [198]:
df_sqli.shape

(248093, 39)

## 194 - Fake the Source of Data

In [199]:
df_fake = df_capec[df_capec['194 - Fake the Source of Data'] == 1].loc[:, ['request_http_method','request_http_request', 'request_body', '194 - Fake the Source of Data']].reset_index(drop=True)

In [200]:
df_fake = df_capec[df_capec['194 - Fake the Source of Data'] == 1]

In [201]:
df_fake.head()

timestamp      src_ip  src_port      dst_ip  dst_port  \
44501  17/Jul/2020:22:24:07 +0200  172.26.0.1   65225.0  172.26.0.4      80.0   
44502  17/Jul/2020:22:24:07 +0200  172.26.0.1   65227.0  172.26.0.4      80.0   
44504  17/Jul/2020:22:24:07 +0200  172.26.0.1   65228.0  172.26.0.4      80.0   
44505  17/Jul/2020:22:24:07 +0200  172.26.0.1   65227.0  172.26.0.4      80.0   
44506  17/Jul/2020:22:24:07 +0200  172.26.0.1   65227.0  172.26.0.4      80.0   

      request_http_method                            request_http_request  \
44501                 GET                 /http%3A%2F%2Fwww.google.com%2F   
44502                 GET       /http%3A%2F%2Fwww.google.com%2F/index.php   
44504                 GET            /http%3A%2F%2Fwww.google.com%3A80%2F   
44505                 GET  /http%3A%2F%2Fwww.google.com%3A80%2F/index.php   
44506                 GET                    /http%3A%2F%2Fwww.google.com   

      request_http_protocol  \
44501              HTTP/1.1   
44502              HTTP/1.1   
44504              HTTP/1.1   
44505              HTTP/1.1   
44506              HTTP/1.1   

                                      request_user_agent request_referer  \
44501  Mozilla/5.0 (X11; Linux x86_64; rv:73.0) Gecko...             NaN   
44502  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...             NaN   
44504  Mozilla/5.0 (X11; Linux x86_64; rv:73.0) Gecko...             NaN   
44505  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...             NaN   
44506  Mozilla/5.0 (X11; Linux x86_64; rv:73.0) Gecko...             NaN   

        request_host request_origin  \
44501  test-site.com            NaN   
44502  test-site.com            NaN   
44504  test-site.com            NaN   
44505  test-site.com            NaN   
44506  test-site.com            NaN   

                                          request_cookie request_content_type  \
44501                                                NaN                  NaN   
44502  wordpress_logged_in_1aefbe2f76edd740f8e362f39d...                  NaN   
44504                                                NaN                  NaN   
44505  wordpress_logged_in_1aefbe2f76edd740f8e362f39d...                  NaN   
44506                                                NaN                  NaN   

                                          request_accept  \
44501  text/html,application/xhtml+xml,application/xm...   
44502                                                NaN   
44504  text/html,application/xhtml+xml,application/xm...   
44505                                                NaN   
44506  text/html,application/xhtml+xml,application/xm...   

      request_accept_language request_accept_encoding  request_do_not_track  \
44501          en-US,en;q=0.5                     NaN                   NaN   
44502                     NaN                     NaN                   NaN   
44504          en-US,en;q=0.5                     NaN                   NaN   
44505                     NaN                     NaN                   NaN   
44506          en-US,en;q=0.5                     NaN                   NaN   

      request_connection request_body response_http_protocol  \
44501         keep-alive          NaN               HTTP/1.1   
44502                NaN          NaN               HTTP/1.1   
44504         keep-alive          NaN               HTTP/1.1   
44505                NaN          NaN               HTTP/1.1   
44506         keep-alive          NaN               HTTP/1.1   

       response_http_status_code response_http_status_message  \
44501                      404.0                    Not Found   
44502                      404.0                    Not Found   
44504                      404.0                    Not Found   
44505                      404.0                    Not Found   
44506                      404.0                    Not Found   

       response_content_length  000 - Normal  272 - Protocol Manipulation  \
44501              

In [202]:
df_fake.shape

(55982, 39)

## 242 - Code Injection

In [203]:
df_injection = df_path_traversal = df_capec[df_capec['242 - Code Injection'] == 1].loc[:, ['request_http_method','request_http_request', 'request_body', '242 - Code Injection']].reset_index(drop=True)

In [204]:
df_injection = df_path_traversal = df_capec[df_capec['242 - Code Injection'] == 1]

In [205]:
df_injection.head()

timestamp      src_ip  src_port      dst_ip  \
7491    17/Jul/2020:20:08:51 +0200  172.26.0.1   36158.0  172.26.0.4   
110025  17/Jul/2020:23:36:11 +0200  172.26.0.1   50100.0  172.26.0.4   
110026  17/Jul/2020:23:36:12 +0200  172.26.0.1   50101.0  172.26.0.4   
110067  17/Jul/2020:23:36:19 +0200  172.26.0.1   50103.0  172.26.0.4   
110068  17/Jul/2020:23:36:19 +0200  172.26.0.1   50101.0  172.26.0.4   

        dst_port request_http_method  \
7491        80.0                 GET   
110025      80.0                POST   
110026      80.0                POST   
110067      80.0                 GET   
110068      80.0                POST   

                                     request_http_request  \
7491    /index.php?s=/Index/\x5Cthink\x5Capp/invokefun...   
110025          /blog/index.php/my-account/edit-password/   
110026          /blog/index.php/my-account/edit-password/   
110067  /blog/index.php/my-account/edit-profile/post.p...   
110068  /blog/index.php/my-account/edit-profile/post.p...   

       request_http_protocol  \
7491                HTTP/1.1   
110025              HTTP/1.1   
110026              HTTP/1.1   
110067              HTTP/1.1   
110068              HTTP/1.1   

                                       request_user_agent  \
7491    Mozilla/5.0 (Windows NT 10.0; Win64; x64) Appl...   
110025  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
110026  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
110067  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
110068  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   

                                          request_referer   request_host  \
7491                                                  NaN  test-site.com   
110025  http://test-site.com/blog/index.php/my-account...  test-site.com   
110026  http://test-site.com/blog/index.php/my-account...  test-site.com   
110067                                                NaN  test-site.com   
110068  http://test-site.com/blog/index.php/my-account...  test-site.com   

       request_origin                         request_cookie  \
7491              NaN                                    NaN   
110025            NaN  wordpress_test_cookie=WP+Cookie+check   
110026            NaN  wordpress_test_cookie=WP+Cookie+check   
110067            NaN  wordpress_test_cookie=WP+Cookie+check   
110068            NaN  wordpress_test_cookie=WP+Cookie+check   

                     request_content_type request_accept  \
7491                                  NaN            */*   
110025  application/x-www-form-urlencoded            NaN   
110026  application/x-www-form-urlencoded            NaN   
110067                                NaN            NaN   
110068  application/x-www-form-urlencoded            NaN   

       request_accept_language request_accept_encoding  request_do_not_track  \
7491                       NaN           deflate, gzip                   NaN   
110025                     NaN                     NaN                   NaN   
110026                     NaN                     NaN                   NaN   
110067                     NaN                     NaN                   NaN   
110068                     NaN                     NaN                   NaN   

       request_connection                                       request_body  \
7491                  NaN                                                NaN   
110025                NaN  username=%22%3E%3Cscript%3Ealert%281%29%3B%3C%...   
110026                NaN  username=%22+onMouseOver%3D%22alert%281%29%3B&...   
110067                NaN                                                NaN   
110068                NaN  username=wakefulViper3&password=GJLU7ABS&user-...   

       response_http_protocol  response_http_status_code  \
7491                 HTTP/1.1                      200.0   
110025               HTTP/1.1                      200.0   
110026               HTTP/1.1                      200.0   
110067             

In [206]:
df_injection.shape

(13792, 39)

## 34 - HTTP Response Splitting

In [207]:
df_response = df_capec[df_capec['34 - HTTP Response Splitting'] == 1].loc[:, ['request_http_method','request_http_request', 'request_body', '34 - HTTP Response Splitting']].reset_index(drop=True)

In [208]:
df_response = df_capec[df_capec['34 - HTTP Response Splitting'] == 1]

In [209]:
df_response.head()

timestamp      src_ip  src_port      dst_ip  \
242072  18/Jul/2020:03:22:32 +0200  172.26.0.1   62915.0  172.26.0.4   
242073  18/Jul/2020:03:22:32 +0200  172.26.0.1   62914.0  172.26.0.4   
242096  18/Jul/2020:03:22:34 +0200  172.26.0.1   62915.0  172.26.0.4   
242097  18/Jul/2020:03:22:34 +0200  172.26.0.1   62914.0  172.26.0.4   
242121  18/Jul/2020:03:22:37 +0200  172.26.0.1   62915.0  172.26.0.4   

        dst_port request_http_method  \
242072      80.0                 GET   
242073      80.0                 GET   
242096      80.0                 GET   
242097      80.0                 GET   
242121      80.0                 GET   

                            request_http_request request_http_protocol  \
242072            /blog%0Acat+%2Fetc%2Fpasswd%0A              HTTP/1.1   
242073  /blog%0Acat+%2Fetc%2Fpasswd%0A/index.php              HTTP/1.1   
242096                       /blog%0Asleep+15%0A              HTTP/1.1   
242097             /blog%0Asleep+15%0A/index.php              HTTP/1.1   
242121  /blog%0Acat+%2Fetc%2Fpasswd%0A/index.php              HTTP/1.1   

                                       request_user_agent request_referer  \
242072  Mozilla/5.0 (X11; Linux x86_64; rv:73.0) Gecko...             NaN   
242073  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...             NaN   
242096  Mozilla/5.0 (X11; Linux x86_64; rv:73.0) Gecko...             NaN   
242097  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...             NaN   
242121  Mozilla/5.0 (X11; Linux x86_64; rv:73.0) Gecko...             NaN   

         request_host request_origin  \
242072  test-site.com            NaN   
242073  test-site.com            NaN   
242096  test-site.com            NaN   
242097  test-site.com            NaN   
242121  test-site.com            NaN   

                                           request_cookie  \
242072                                                NaN   
242073  wordpress_logged_in_1aefbe2f76edd740f8e362f39d...   
242096                                                NaN   
242097  wordpress_logged_in_1aefbe2f76edd740f8e362f39d...   
242121                                                NaN   

       request_content_type  \
242072                  NaN   
242073                  NaN   
242096                  NaN   
242097                  NaN   
242121                  NaN   

                                           request_accept  \
242072  text/html,application/xhtml+xml,application/xm...   
242073                                                NaN   
242096  text/html,application/xhtml+xml,application/xm...   
242097                                                NaN   
242121  text/html,application/xhtml+xml,application/xm...   

       request_accept_language request_accept_encoding  request_do_not_track  \
242072          en-US,en;q=0.5                     NaN                   NaN   
242073                     NaN                     NaN                   NaN   
242096          en-US,en;q=0.5                     NaN                   NaN   
242097                     NaN                     NaN                   NaN   
242121          en-US,en;q=0.5                     NaN                   NaN   

       request_connection request_body response_http_protocol  \
242072         keep-alive          NaN               HTTP/1.1   
242073                NaN          NaN               HTTP/1.1   
242096         keep-alive          NaN               HTTP/1.1   
242097                NaN          NaN               HTTP/1.1   
242121         keep-alive          NaN               HTTP/1.1   

        response_http_status_code response_http_status_message  \
242072                      404.0                    Not Found   
242073                      404.0                    Not Found   
242096                      404.0                    Not Found   
242097                      404.0                    Not Found   
242121                      404.0                    Not Found   

        resp

In [210]:
df_response.shape

(19134, 39)

## 000 - Normal

In [211]:
df_normal = df_capec[df_capec['000 - Normal'] == 1].loc[:, ['request_http_method','request_http_request', 'request_body', '000 - Normal']].reset_index(drop=True)

In [212]:
df_normal = df_capec[df_capec['000 - Normal'] == 1]

In [213]:
df_normal.head()

timestamp      src_ip  src_port      dst_ip  dst_port  \
1  17/Jul/2020:12:23:41 +0100  172.26.0.1   55897.0  172.26.0.4      80.0   
2  17/Jul/2020:12:23:51 +0100  172.26.0.1   55901.0  172.26.0.4      80.0   
4  17/Jul/2020:12:23:58 +0100  172.26.0.1   55903.0  172.26.0.4      80.0   
5  17/Jul/2020:12:24:07 +0100  172.26.0.1   55910.0  172.26.0.4      80.0   
6  17/Jul/2020:12:24:15 +0100  172.26.0.1   55911.0  172.26.0.4      80.0   

  request_http_method                               request_http_request  \
1                 GET  /blog/index.php/2020/04/04/voluptatum-reprehen...   
2                 GET                               /blog/xmlrpc.php?rsd   
4                 GET  /blog/index.php/2020/04/04/nihil-tenetur-et-ad...   
5                 GET  /blog/index.php/2020/04/04/explicabo-qui-fuga-...   
6                 GET  /blog/index.php/2020/04/04/explicabo-qui-fuga-...   

  request_http_protocol                                 request_user_agent  \
1              HTTP/1.1  Mozilla/5.0 (Macintosh; Intel Mac OS X 10.6; r...   
2              HTTP/1.1  Mozilla/4.0 (compatible; MSIE 6.0; Windows NT ...   
4              HTTP/1.1  Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_2...   
5              HTTP/1.1  Mozilla/5.0 (Windows NT 10.0; Win64; x64) Appl...   
6              HTTP/1.1  Mozilla/5.0 (Windows NT 5.1) AppleWebKit/537.3...   

  request_referer   request_host request_origin request_cookie  \
1             NaN  test-site.com            NaN            NaN   
2             NaN  test-site.com            NaN            NaN   
4             NaN  test-site.com            NaN            NaN   
5             NaN  test-site.com            NaN            NaN   
6             NaN  test-site.com            NaN            NaN   

  request_content_type request_accept request_accept_language  \
1                  NaN            */*                     NaN   
2                  NaN            */*                     NaN   
4                  NaN            */*                     NaN   
5                  NaN            */*                     NaN   
6                  NaN            */*                     NaN   

  request_accept_encoding  request_do_not_track request_connection  \
1           gzip, deflate                   NaN         keep-alive   
2           gzip, deflate                   NaN         keep-alive   
4           gzip, deflate                   NaN         keep-alive   
5           gzip, deflate                   NaN         keep-alive   
6           gzip, deflate                   NaN         keep-alive   

  request_body response_http_protocol  response_http_status_code  \
1          NaN               HTTP/1.1                      200.0   
2          NaN               HTTP/1.1                      200.0   
4          NaN               HTTP/1.1                      200.0   
5          NaN               HTTP/1.1                      200.0   
6          NaN               HTTP/1.1                      200.0   

  response_http_status_message  response_content_length  000 - Normal  \
1                           OK                  20862.0             1   
2                           OK                    317.0             1   
4                           OK                  20863.0             1   
5                           OK                  21062.0             1   
6                           OK                  21061.0             1   

   272 - Protocol Manipulation  242 - Code Injection  \
1                            0                     0   
2                            0                     0   
4                            0                     0   
5                            0                     0   
6                            0                     0   

   88 - OS Command Injection  126 - Path Traversal  66 - SQL Injection  \
1                          0                     0                   0   
2                          0                     0                   0   
4              

In [214]:
df_normal.shape

(226509, 39)

## 272 - Protocol Manipulation

In [215]:
df_protocol = df_capec[df_capec['272 - Protocol Manipulation'] == 1].loc[:, ['request_http_method','request_http_request', 'request_body', '000 - Normal']].reset_index(drop=True)

In [216]:
df_protocol = df_capec[df_capec['272 - Protocol Manipulation'] == 1]

In [217]:
df_protocol.head()

timestamp      src_ip  src_port      dst_ip  dst_port  \
3461  17/Jul/2020:19:30:37 +0200  172.26.0.1   59927.0  172.26.0.4      80.0   
7138  17/Jul/2020:19:58:39 +0200  172.26.0.1   63015.0  172.26.0.4      80.0   
7139  17/Jul/2020:19:58:39 +0200  172.26.0.1   63016.0  172.26.0.4      80.0   
7508  17/Jul/2020:20:09:38 +0200  172.26.0.1   64214.0  172.26.0.4      80.0   
7509  17/Jul/2020:20:09:38 +0200  172.26.0.1   64213.0  172.26.0.4      80.0   

     request_http_method                               request_http_request  \
3461                 GET                        /blog/index.php/my-account/   
7138                 GET  /blog/wp-content/uploads/2020/04/c874f043-4188...   
7139                 GET  /blog/wp-content/uploads/2020/04/5d0beb80-cdb9...   
7508                 GET  /blog/wp-content/uploads/2020/04/8427d570-74b5...   
7509                 GET  /blog/wp-content/uploads/2020/04/cda5c6a3-e3d9...   

     request_http_protocol                                 request_user_agent  \
3461              HTTP/1.1  Mozilla/5.0 (X11; Linux x86_64; rv:73.0) Gecko...   
7138              HTTP/1.1  Mozilla/5.0 (X11; Linux x86_64; rv:73.0) Gecko...   
7139              HTTP/1.1  Mozilla/5.0 (X11; Linux x86_64; rv:73.0) Gecko...   
7508              HTTP/1.1  Mozilla/5.0 (X11; Linux x86_64; rv:73.0) Gecko...   
7509              HTTP/1.1  Mozilla/5.0 (X11; Linux x86_64; rv:73.0) Gecko...   

            request_referer request_host request_origin request_cookie  \
3461                    NaN   172.26.0.1            NaN            NaN   
7138                    NaN   172.26.0.1            NaN            NaN   
7139                    NaN   172.26.0.1            NaN            NaN   
7508  http://test-site.com/   172.26.0.1            NaN            NaN   
7509  http://test-site.com/   172.26.0.1            NaN            NaN   

     request_content_type                                     request_accept  \
3461                  NaN  text/html,application/xhtml+xml,application/xm...   
7138                  NaN                                     image/webp,*/*   
7139                  NaN                                     image/webp,*/*   
7508                  NaN                                     image/webp,*/*   
7509                  NaN                                     image/webp,*/*   

     request_accept_language request_accept_encoding  request_do_not_track  \
3461          en-US,en;q=0.5                     NaN                   1.0   
7138          en-US,en;q=0.5                     NaN                   1.0   
7139          en-US,en;q=0.5                     NaN                   1.0   
7508          en-US,en;q=0.5           gzip, deflate                   1.0   
7509          en-US,en;q=0.5           gzip, deflate                   1.0   

     request_connection request_body response_http_protocol  \
3461         keep-alive          NaN               HTTP/1.1   
7138         keep-alive          NaN               HTTP/1.1   
7139         keep-alive          NaN               HTTP/1.1   
7508         keep-alive          NaN               HTTP/1.1   
7509         keep-alive          NaN               HTTP/1.1   

      response_http_status_code response_http_status_message  \
3461                      200.0                           OK   
7138                      200.0                           OK   
7139                      200.0                           OK   
7508                      200.0                           OK   
7509                      200.0                           OK   

      response_content_length  000 - Normal  272 - Protocol Manipulation  \
3461                  57822.0             0                            1   
7138                  78796.0             0                            1   
7139                 236440.0             0                            1   
7508                   2255.0             0                            1   
7509                   1461.0            

In [218]:
df_protocol.shape

(6924, 39)

## 88 - OS Command Injection

In [219]:
df_os = df_capec[df_capec['88 - OS Command Injection'] == 1].loc[:, ['request_http_method','request_http_request', 'request_body', '000 - Normal']].reset_index(drop=True)

In [220]:
df_os = df_capec[df_capec['88 - OS Command Injection'] == 1]

In [221]:
df_os.head()

timestamp      src_ip  src_port      dst_ip  \
7494    17/Jul/2020:20:08:52 +0200  172.26.0.1   36158.0  172.26.0.4   
257238  18/Jul/2020:04:08:39 +0200  172.26.0.1   63219.0  172.26.0.4   
257240  18/Jul/2020:04:08:39 +0200  172.26.0.1   63219.0  172.26.0.4   
257242  18/Jul/2020:04:08:40 +0200  172.26.0.1   63219.0  172.26.0.4   
257244  18/Jul/2020:04:08:40 +0200  172.26.0.1   63219.0  172.26.0.4   

        dst_port request_http_method  \
7494        80.0                 GET   
257238      80.0                POST   
257240      80.0                POST   
257242      80.0                POST   
257244      80.0                POST   

                                     request_http_request  \
7494    /index.php?s=/module/action/param1/${@die(sha1...   
257238          /blog/index.php/my-account/edit-password/   
257240          /blog/index.php/my-account/edit-password/   
257242          /blog/index.php/my-account/edit-password/   
257244          /blog/index.php/my-account/edit-password/   

       request_http_protocol  \
7494                HTTP/1.1   
257238              HTTP/1.1   
257240              HTTP/1.1   
257242              HTTP/1.1   
257244              HTTP/1.1   

                                       request_user_agent  \
7494    Mozilla/5.0 (Windows NT 10.0; Win64; x64) Appl...   
257238  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
257240  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
257242  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
257244  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   

                                          request_referer   request_host  \
7494                                                  NaN  test-site.com   
257238  http://test-site.com/blog/index.php/my-account...  test-site.com   
257240  http://test-site.com/blog/index.php/my-account...  test-site.com   
257242  http://test-site.com/blog/index.php/my-account...  test-site.com   
257244  http://test-site.com/blog/index.php/my-account...  test-site.com   

       request_origin                                     request_cookie  \
7494              NaN                                                NaN   
257238            NaN  comment_author_url_1aefbe2f76edd740f8e362f39da...   
257240            NaN  comment_author_url_1aefbe2f76edd740f8e362f39da...   
257242            NaN  comment_author_url_1aefbe2f76edd740f8e362f39da...   
257244            NaN  comment_author_url_1aefbe2f76edd740f8e362f39da...   

                     request_content_type request_accept  \
7494                                  NaN            */*   
257238  application/x-www-form-urlencoded            NaN   
257240  application/x-www-form-urlencoded            NaN   
257242  application/x-www-form-urlencoded            NaN   
257244  application/x-www-form-urlencoded            NaN   

       request_accept_language request_accept_encoding  request_do_not_track  \
7494                       NaN           deflate, gzip                   NaN   
257238                     NaN                     NaN                   NaN   
257240                     NaN                     NaN                   NaN   
257242                     NaN                     NaN                   NaN   
257244                     NaN                     NaN                   NaN   

       request_connection                                       request_body  \
7494                  NaN                                                NaN   
257238                NaN  password_current=aH2q*Rj%26sleep+15%26&passwor...   
257240                NaN  password_current=H#bO2%3Bsleep+15%3B&password_...   
257242                NaN  password_current=muRWD%22%26sleep+15%26%22&pas...   
257244                NaN  password_current=3X*Bam%22%3Bsleep+15%3B%22&pa...   

       response_http_protocol  response_http_status_code  \
7494                 HTTP/1.1                      200.0   
257238               HTTP/1.1                      200.0   
257240 

In [222]:
df_os.shape

(3074, 39)

## 126 - Path Traversal

In [223]:
df_path_traveral = df_capec[df_capec['126 - Path Traversal'] == 1].loc[:, ['request_http_method','request_http_request', 'request_body', '000 - Normal']].reset_index(drop=True)

In [224]:
df_path_traveral = df_capec[df_capec['126 - Path Traversal'] == 1]

In [225]:
df_path_traveral.head()

timestamp      src_ip  src_port      dst_ip  dst_port  \
2968  17/Jul/2020:19:25:15 +0200  172.26.0.1   59862.0  172.26.0.4      80.0   
2970  17/Jul/2020:19:25:15 +0200  172.26.0.1   59862.0  172.26.0.4      80.0   
2973  17/Jul/2020:19:25:16 +0200  172.26.0.1   59862.0  172.26.0.4      80.0   
2975  17/Jul/2020:19:25:16 +0200  172.26.0.1   59862.0  172.26.0.4      80.0   
2976  17/Jul/2020:19:25:16 +0200  172.26.0.1   59862.0  172.26.0.4      80.0   

     request_http_method                     request_http_request  \
2968                 GET                            /.svn/entries   
2970                 GET                              /.git/index   
2973                 GET                       /blog/.svn/entries   
2975                 GET                         /blog/.git/index   
2976                 GET  /blog/index.php/my-account/.svn/entries   

     request_http_protocol                                 request_user_agent  \
2968              HTTP/1.1  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
2970              HTTP/1.1  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
2973              HTTP/1.1  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
2975              HTTP/1.1  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
2976              HTTP/1.1  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   

     request_referer   request_host request_origin  \
2968             NaN  test-site.com            NaN   
2970             NaN  test-site.com            NaN   
2973             NaN  test-site.com            NaN   
2975             NaN  test-site.com            NaN   
2976             NaN  test-site.com            NaN   

                                         request_cookie request_content_type  \
2968                                                NaN                  NaN   
2970                                                NaN                  NaN   
2973  wordpress_logged_in_1aefbe2f76edd740f8e362f39d...                  NaN   
2975  wordpress_logged_in_1aefbe2f76edd740f8e362f39d...                  NaN   
2976  wordpress_logged_in_1aefbe2f76edd740f8e362f39d...                  NaN   

     request_accept request_accept_language request_accept_encoding  \
2968            NaN                     NaN                     NaN   
2970            NaN                     NaN                     NaN   
2973            NaN                     NaN                     NaN   
2975            NaN                     NaN                     NaN   
2976            NaN                     NaN                     NaN   

      request_do_not_track request_connection request_body  \
2968                   NaN                NaN          NaN   
2970                   NaN                NaN          NaN   
2973                   NaN                NaN          NaN   
2975                   NaN                NaN          NaN   
2976                   NaN                NaN          NaN   

     response_http_protocol  response_http_status_code  \
2968               HTTP/1.1                      404.0   
2970               HTTP/1.1                      404.0   
2973               HTTP/1.1                      404.0   
2975               HTTP/1.1                      404.0   
2976               HTTP/1.1                      404.0   

     response_http_status_message  response_content_length  000 - Normal  \
2968                    Not Found                    276.0             0   
2970                    Not Found                    276.0             0   
2973                    Not Found                    276.0             0   
2975                    Not Found                    276.0             0   
2976                    Not Found                  58226.0             0   

      272 - Protocol Manipulation  242 - Code Injection  \
2968                            0                     0   
2970                            0                     0   
2973                            0                  

In [226]:
df_path_traveral.shape

(17595, 39)

## 16 - Dictionary-based Password Attack

In [227]:
df_dict = df_capec[df_capec['16 - Dictionary-based Password Attack'] == 1]

In [228]:
df_dict.head()

timestamp      src_ip  src_port      dst_ip  dst_port  \
3122  17/Jul/2020:19:25:48 +0200  172.26.0.1   59863.0  172.26.0.4      80.0   
3351  17/Jul/2020:19:26:25 +0200  172.26.0.1   59870.0  172.26.0.4      80.0   
4565  17/Jul/2020:19:33:33 +0200  172.26.0.1   59973.0  172.26.0.4      80.0   
4730  17/Jul/2020:19:33:50 +0200  172.26.0.1   59975.0  172.26.0.4      80.0   
4731  17/Jul/2020:19:33:50 +0200  172.26.0.1   59975.0  172.26.0.4      80.0   

     request_http_method request_http_request request_http_protocol  \
3122                POST   /blog/wp-login.php              HTTP/1.1   
3351                POST   /blog/wp-login.php              HTTP/1.1   
4565                POST   /blog/wp-login.php              HTTP/1.1   
4730                POST   /blog/wp-login.php              HTTP/1.1   
4731                POST   /blog/wp-login.php              HTTP/1.1   

                                     request_user_agent  \
3122  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
3351  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
4565  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
4730  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
4731  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   

                                        request_referer   request_host  \
3122             http://test-site.com/blog/wp-login.php  test-site.com   
3351  http://test-site.com/blog/wp-login.php?loggedo...  test-site.com   
4565             http://test-site.com/blog/wp-login.php  test-site.com   
4730  http://test-site.com/blog/wp-login.php?reauth=...  test-site.com   
4731  http://test-site.com/blog/wp-login.php?reauth=...  test-site.com   

     request_origin                         request_cookie  \
3122            NaN  wordpress_test_cookie=WP+Cookie+check   
3351            NaN  wordpress_test_cookie=WP+Cookie+check   
4565            NaN  wordpress_test_cookie=WP+Cookie+check   
4730            NaN  wordpress_test_cookie=WP+Cookie+check   
4731            NaN  wordpress_test_cookie=WP+Cookie+check   

                   request_content_type request_accept  \
3122  application/x-www-form-urlencoded            NaN   
3351  application/x-www-form-urlencoded            NaN   
4565  application/x-www-form-urlencoded            NaN   
4730  application/x-www-form-urlencoded            NaN   
4731  application/x-www-form-urlencoded            NaN   

     request_accept_language request_accept_encoding  request_do_not_track  \
3122                     NaN                     NaN                   NaN   
3351                     NaN                     NaN                   NaN   
4565                     NaN                     NaN                   NaN   
4730                     NaN                     NaN                   NaN   
4731                     NaN                     NaN                   NaN   

     request_connection                                       request_body  \
3122                NaN  log=rafael&pwd=975cU&rememberme=forever&wp-sub...   
3351                NaN  log=thrilledCrackers6&pwd=hOKxTw_2&rememberme=...   
4565                NaN  log=innocentOwl1&pwd=41IPdtwh&rememberme=forev...   
4730                NaN  log=panickyEagle4&pwd=cXTkX&rememberme=forever...   
4731                NaN  log=trustingOatmeal8&pwd=qqIzM7vt&rememberme=f...   

     response_http_protocol  response_http_status_code  \
3122               HTTP/1.1                      200.0   
3351               HTTP/1.1                      200.0   
4565               HTTP/1.1                      200.0   
4730               HTTP/1.1                      200.0   
4731               HTTP/1.1                      200.0   

     response_http_status_message  response_content_length  000 - Normal  \
3122                           OK                   4215.0             0   
3351                           OK                   4209.0             0   
4565                           OK                   4209.0             0   
47

In [229]:
df_dict.shape

(836, 39)

## 310 - Scanning for Vulnerable Software

In [230]:
df_scanning = df_capec[df_capec['310 - Scanning for Vulnerable Software'] == 1]

In [231]:
df_scanning.tail()

timestamp      src_ip  src_port      dst_ip  dst_port  \
7473  17/Jul/2020:20:06:19 +0200  172.26.0.1   64144.0  172.26.0.4      80.0   
7480  17/Jul/2020:20:06:21 +0200  172.26.0.1   64151.0  172.26.0.4      80.0   
7481  17/Jul/2020:20:06:21 +0200  172.26.0.1   64152.0  172.26.0.4      80.0   
7482  17/Jul/2020:20:06:21 +0200  172.26.0.1   64153.0  172.26.0.4      80.0   
7483  17/Jul/2020:20:06:21 +0200  172.26.0.1   64154.0  172.26.0.4      80.0   

     request_http_method           request_http_request request_http_protocol  \
7473                 GET  /mt-static/mt-config.cgi.save              HTTP/1.1   
7480                 GET            /%23settings.php%23              HTTP/1.1   
7481                 GET           /settings%20copy.php              HTTP/1.1   
7482                 GET      /Copy%20of%20settings.php              HTTP/1.1   
7483                 GET             /settings.php.save              HTTP/1.1   

                                     request_user_agent request_referer  \
7473  Mozilla/5.0 (compatible; Nmap Scripting Engine...             NaN   
7480  Mozilla/5.0 (compatible; Nmap Scripting Engine...             NaN   
7481  Mozilla/5.0 (compatible; Nmap Scripting Engine...             NaN   
7482  Mozilla/5.0 (compatible; Nmap Scripting Engine...             NaN   
7483  Mozilla/5.0 (compatible; Nmap Scripting Engine...             NaN   

                                    request_host request_origin  \
7473  249.network5-175-62-15.static.network5.net            NaN   
7480  249.network5-175-62-15.static.network5.net            NaN   
7481  249.network5-175-62-15.static.network5.net            NaN   
7482  249.network5-175-62-15.static.network5.net            NaN   
7483  249.network5-175-62-15.static.network5.net            NaN   

     request_cookie request_content_type request_accept  \
7473            NaN                  NaN            NaN   
7480            NaN                  NaN            NaN   
7481            NaN                  NaN            NaN   
7482            NaN                  NaN            NaN   
7483            NaN                  NaN            NaN   

     request_accept_language request_accept_encoding  request_do_not_track  \
7473                     NaN                     NaN                   NaN   
7480                     NaN                     NaN                   NaN   
7481                     NaN                     NaN                   NaN   
7482                     NaN                     NaN                   NaN   
7483                     NaN                     NaN                   NaN   

     request_connection request_body response_http_protocol  \
7473              close          NaN               HTTP/1.1   
7480              close          NaN               HTTP/1.1   
7481              close          NaN               HTTP/1.1   
7482              close          NaN               HTTP/1.1   
7483              close          NaN               HTTP/1.1   

      response_http_status_code response_http_status_message  \
7473                      404.0                    Not Found   
7480                      404.0                    Not Found   
7481                      404.0                    Not Found   
7482                      404.0                    Not Found   
7483                      404.0                    Not Found   

      response_content_length  000 - Normal  272 - Protocol Manipulation  \
7473                    301.0             0                            0   
7480                    301.0             0                            0   
7481                    301.0             0                            0   
7482                    301.0             0                            0   
7483                    301.0             0                            0   

      242 - Code Injection  88 - OS Command Injection  126 - Path Traversal  \
7473                     0                          0                     0  

In [232]:
df_scanning.shape

(2382, 39)

## 153 - Input Data Manipulation

In [233]:
df_input = df_capec[df_capec['153 - Input Data Manipulation'] == 1]

In [234]:
df_input.head()

timestamp      src_ip  src_port      dst_ip  dst_port  \
19406  17/Jul/2020:21:48:16 +0200  172.26.0.1   64765.0  172.26.0.4      80.0   
19411  17/Jul/2020:21:48:17 +0200  172.26.0.1   64763.0  172.26.0.4      80.0   
19413  17/Jul/2020:21:48:17 +0200  172.26.0.1   64765.0  172.26.0.4      80.0   
19415  17/Jul/2020:21:48:17 +0200  172.26.0.1   64765.0  172.26.0.4      80.0   
19416  17/Jul/2020:21:48:17 +0200  172.26.0.1   64763.0  172.26.0.4      80.0   

      request_http_method                       request_http_request  \
19406                POST  /blog/index.php/my-account/edit-password/   
19411                POST  /blog/index.php/my-account/edit-password/   
19413                POST  /blog/index.php/my-account/edit-password/   
19415                POST  /blog/index.php/my-account/edit-password/   
19416                POST  /blog/index.php/my-account/edit-password/   

      request_http_protocol  \
19406              HTTP/1.1   
19411              HTTP/1.1   
19413              HTTP/1.1   
19415              HTTP/1.1   
19416              HTTP/1.1   

                                      request_user_agent  \
19406  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
19411  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
19413  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
19415  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
19416  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   

                                         request_referer   request_host  \
19406  http://test-site.com/blog/index.php/my-account...  test-site.com   
19411  http://test-site.com/blog/index.php/my-account...  test-site.com   
19413  http://test-site.com/blog/index.php/my-account...  test-site.com   
19415  http://test-site.com/blog/index.php/my-account...  test-site.com   
19416  http://test-site.com/blog/index.php/my-account...  test-site.com   

      request_origin                         request_cookie  \
19406            NaN  wordpress_test_cookie=WP+Cookie+check   
19411            NaN  wordpress_test_cookie=WP+Cookie+check   
19413            NaN  wordpress_test_cookie=WP+Cookie+check   
19415            NaN  wordpress_test_cookie=WP+Cookie+check   
19416            NaN  wordpress_test_cookie=WP+Cookie+check   

      request_content_type request_accept request_accept_language  \
19406          /etc/passwd            NaN                     NaN   
19411           etc/passwd            NaN                     NaN   
19413                  c:/            NaN                     NaN   
19415                    /            NaN                     NaN   
19416                  c:\            NaN                     NaN   

      request_accept_encoding  request_do_not_track request_connection  \
19406                     NaN                   NaN                NaN   
19411                     NaN                   NaN                NaN   
19413                     NaN                   NaN                NaN   
19415                     NaN                   NaN                NaN   
19416                     NaN                   NaN                NaN   

                                            request_body  \
19406  password_current=76U8B7&password_1=76U8B7&pass...   
19411  password_current=ltHgBOO&password_1=ltHgBOO&pa...   
19413  password_current=5uSt9Xd&password_1=5uSt9Xd&pa...   
19415  password_current=J875f&password_1=J875f&passwo...   
19416  password_current=C&rclFh&password_1=C&rclFh&pa...   

      response_http_protocol  response_http_status_code  \
19406               HTTP/1.1                      200.0   
19411               HTTP/1.1                      200.0   
19413               HTTP/1.1                      200.0   
19415               HTTP/1.1                      200.0   
19416               HTTP/1.1                      200.0   

      response_http_status_message  response_content_length  000 - Normal  \
19406                           OK                  57821.0             0

In [235]:
df_input.shape

(1387, 39)

## 274 - HTTP Verb Tampering

In [236]:
df_http_verb = df_capec[df_capec['274 - HTTP Verb Tampering'] == 1]

In [237]:
df_http_verb.tail()

timestamp      src_ip  src_port      dst_ip  \
543488  19/Jul/2020:11:49:29 +0200  172.26.0.1   56039.0  172.26.0.4   
543489  19/Jul/2020:11:49:29 +0200  172.26.0.1   56039.0  172.26.0.4   
543490  19/Jul/2020:11:49:29 +0200  172.26.0.1   56039.0  172.26.0.4   
543491  19/Jul/2020:11:49:29 +0200  172.26.0.1   56039.0  172.26.0.4   
543492  19/Jul/2020:11:49:29 +0200  172.26.0.1   56039.0  172.26.0.4   

        dst_port request_http_method request_http_request  \
543488      80.0               MKCOL         /sitemap.xml   
543489      80.0                MOVE         /sitemap.xml   
543490      80.0            PROPFIND         /sitemap.xml   
543491      80.0           PROPPATCH         /sitemap.xml   
543492      80.0              UNLOCK         /sitemap.xml   

       request_http_protocol  \
543488              HTTP/1.1   
543489              HTTP/1.1   
543490              HTTP/1.1   
543491              HTTP/1.1   
543492              HTTP/1.1   

                                       request_user_agent request_referer  \
543488  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...             NaN   
543489  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...             NaN   
543490  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...             NaN   
543491  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...             NaN   
543492  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...             NaN   

         request_host request_origin request_cookie request_content_type  \
543488  test-site.com            NaN            NaN                  NaN   
543489  test-site.com            NaN            NaN                  NaN   
543490  test-site.com            NaN            NaN                  NaN   
543491  test-site.com            NaN            NaN                  NaN   
543492  test-site.com            NaN            NaN                  NaN   

       request_accept request_accept_language request_accept_encoding  \
543488            NaN                     NaN                     NaN   
543489            NaN                     NaN                     NaN   
543490            NaN                     NaN                     NaN   
543491            NaN                     NaN                     NaN   
543492            NaN                     NaN                     NaN   

        request_do_not_track request_connection request_body  \
543488                   NaN                NaN          NaN   
543489                   NaN                NaN          NaN   
543490                   NaN                NaN          NaN   
543491                   NaN                NaN          NaN   
543492                   NaN                NaN          NaN   

       response_http_protocol  response_http_status_code  \
543488               HTTP/1.1                      405.0   
543489               HTTP/1.1                      405.0   
543490               HTTP/1.1                      405.0   
543491               HTTP/1.1                      405.0   
543492               HTTP/1.1                      405.0   

       response_http_status_message  response_content_length  000 - Normal  \
543488           Method Not Allowed                    302.0             0   
543489           Method Not Allowed                    301.0             0   
543490           Method Not Allowed                    305.0             0   
543491           Method Not Allowed                    306.0             0   
543492           Method Not Allowed                    303.0             0   

        272 - Protocol Manipulation  242 - Code Injection  \
543488                            0                     0   
543489                            0                     0   
543490                            0                     0   
543491                            0                     0   
543492                            0                     0   

        88 - OS Command Injection  126 - Path Traversal  66 - SQL Injection  \
543488      

In [238]:
df_http_verb.shape

(4047, 39)

## Note on data
- Có 2 trường attack ko có tấn công (33 - HTTP Request Smuggling và 248 - Command Injection) trong bộ dataset multiclass. Lý do có thể vì hai trường attack này giao với các trường attack khác (multilabel)
- Tổng cộng các loại tấn công (đã bỏ 2 loại attack bên trên): 12 (bao gồm normal)
    1. 000 - Normal: 226509
    2. 272 - Protocol Manipulation: 6924 
    3. 242 - Code Injection: 13792
    4. 88 - OS Command Injection: 3074
    5. 126 - Path Traversal: 17595 
    6. 66 - SQL Injection: 248093
    7. 16 - Dictionary-based Password Attack: 836 X
    8. 310 - Scanning for Vulnerable Software: 2382 
    9. 153 - Input Data Manipulation: 1387
    10. 274 - HTTP Verb Tampering: 4047 X
    11. 194 - Fake the Source of Data: 55982
    12. 34 - HTTP Response Splitting: 19134 
- Mỗi sample có rất nhiều trường dữ liệu, tuy nhiên chỉ có 1 số attribute giúp classify web attacking
- Thử nghiệm với concat 2 trường dữ liệu: request_http_request và request_body (với method POST)
- Thử nghiệm với một vài kiểu tấn công, và để lại một vài kiểu tấn công có cùng variance với bộ dataset để nghiên cứu Transfer Learning
- Các kiểu tấn công sử dụng trong dataset: 272 - Protocol Manipulation, 242 - Code Injection, 126 - Path Traversal, 66 - SQL Injection, 310 - Scanning for Vulnerable Software, 153 - Input Data Manipulation, 194   - Fake the Source of Data, 34 - HTTP Response Splitting
- Kiểu tấn công dành cho Transfer Learning: OS Command Injection
- Hiện tại đang chưa tích hợp bộ dataset CSIC 2010-2012 (Bao gồm XSS attack và SQLi attack)

## Prepare and format data

In [239]:
# Concat http_method with http_request and http_content

In [240]:
def create_text(method, request, body):
    if method == 'GET':
        return str(method) + ' ' + str(request)
    return str(method) + ' ' + str(request) + ' ' + str(body)
            
df_capec['text'] = df_capec.apply(lambda x: create_text(x['request_http_method'], x['request_http_request'], x['request_body']), axis = 1)

In [241]:
df_capec.head()

timestamp      src_ip  src_port      dst_ip  dst_port  \
1  17/Jul/2020:12:23:41 +0100  172.26.0.1   55897.0  172.26.0.4      80.0   
2  17/Jul/2020:12:23:51 +0100  172.26.0.1   55901.0  172.26.0.4      80.0   
4  17/Jul/2020:12:23:58 +0100  172.26.0.1   55903.0  172.26.0.4      80.0   
5  17/Jul/2020:12:24:07 +0100  172.26.0.1   55910.0  172.26.0.4      80.0   
6  17/Jul/2020:12:24:15 +0100  172.26.0.1   55911.0  172.26.0.4      80.0   

  request_http_method                               request_http_request  \
1                 GET  /blog/index.php/2020/04/04/voluptatum-reprehen...   
2                 GET                               /blog/xmlrpc.php?rsd   
4                 GET  /blog/index.php/2020/04/04/nihil-tenetur-et-ad...   
5                 GET  /blog/index.php/2020/04/04/explicabo-qui-fuga-...   
6                 GET  /blog/index.php/2020/04/04/explicabo-qui-fuga-...   

  request_http_protocol                                 request_user_agent  \
1              HTTP/1.1  Mozilla/5.0 (Macintosh; Intel Mac OS X 10.6; r...   
2              HTTP/1.1  Mozilla/4.0 (compatible; MSIE 6.0; Windows NT ...   
4              HTTP/1.1  Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_2...   
5              HTTP/1.1  Mozilla/5.0 (Windows NT 10.0; Win64; x64) Appl...   
6              HTTP/1.1  Mozilla/5.0 (Windows NT 5.1) AppleWebKit/537.3...   

  request_referer   request_host request_origin request_cookie  \
1             NaN  test-site.com            NaN            NaN   
2             NaN  test-site.com            NaN            NaN   
4             NaN  test-site.com            NaN            NaN   
5             NaN  test-site.com            NaN            NaN   
6             NaN  test-site.com            NaN            NaN   

  request_content_type request_accept request_accept_language  \
1                  NaN            */*                     NaN   
2                  NaN            */*                     NaN   
4                  NaN            */*                     NaN   
5                  NaN            */*                     NaN   
6                  NaN            */*                     NaN   

  request_accept_encoding  request_do_not_track request_connection  \
1           gzip, deflate                   NaN         keep-alive   
2           gzip, deflate                   NaN         keep-alive   
4           gzip, deflate                   NaN         keep-alive   
5           gzip, deflate                   NaN         keep-alive   
6           gzip, deflate                   NaN         keep-alive   

  request_body response_http_protocol  response_http_status_code  \
1          NaN               HTTP/1.1                      200.0   
2          NaN               HTTP/1.1                      200.0   
4          NaN               HTTP/1.1                      200.0   
5          NaN               HTTP/1.1                      200.0   
6          NaN               HTTP/1.1                      200.0   

  response_http_status_message  response_content_length  000 - Normal  \
1                           OK                  20862.0             1   
2                           OK                    317.0             1   
4                           OK                  20863.0             1   
5                           OK                  21062.0             1   
6                           OK                  21061.0             1   

   272 - Protocol Manipulation  242 - Code Injection  \
1                            0                     0   
2                            0                     0   
4                            0                     0   
5                            0                     0   
6                            0                     0   

   88 - OS Command Injection  126 - Path Traversal  66 - SQL Injection  \
1                          0                     0                   0   
2                          0                     0                   0   
4              

In [243]:
# Mapping values for classification

In [247]:
def count_duplicate(a0,a1,a2,a3,a4,a5,a6,a7,a8,a9,a10,a11,a12,a13):
    if a0 == 1:
        return '000 - Normal'
    elif a1 == 1:
        return '272 - Protocol Manipulation'
    elif a2 == 1:
        return '242 - Code Injection'
    elif a3 == 1:
        return '88 - OS Command Injection'
    elif a4 == 1:
        return '126 - Path Traversal'
    elif a5 == 1:
        return '66 - SQL Injection'
    elif a6 == 1:
        return '16 - Dictionary-based Password Attack'
    elif a7 == 1:
        return '310 - Scanning for Vulnerable Software'
    elif a8 == 1:
        return '153 - Input Data Manipulation'
    elif a9 == 1:
        return '248 - Command Injection'
    elif a10 == 1:
        return '274 - HTTP Verb Tampering'
    elif a11 == 1:
        return '194 - Fake the Source of Data'
    elif a12 == 1:
        return '34 - HTTP Response Splitting'
    elif a13 == 1:
        return '33 - HTTP Request Smuggling'
    return 'Error'

        
df_capec['label'] = df_capec.apply(lambda x: count_duplicate(x['000 - Normal'],x['272 - Protocol Manipulation'], x['242 - Code Injection'],x['88 - OS Command Injection'],x['126 - Path Traversal'],x['66 - SQL Injection'],x['16 - Dictionary-based Password Attack'],x['310 - Scanning for Vulnerable Software'],x['153 - Input Data Manipulation'],x['248 - Command Injection'],x['274 - HTTP Verb Tampering'],x['194 - Fake the Source of Data'],x['34 - HTTP Response Splitting'],x['33 - HTTP Request Smuggling']), axis = 1)

In [248]:
df_capec.head()

timestamp      src_ip  src_port      dst_ip  dst_port  \
1  17/Jul/2020:12:23:41 +0100  172.26.0.1   55897.0  172.26.0.4      80.0   
2  17/Jul/2020:12:23:51 +0100  172.26.0.1   55901.0  172.26.0.4      80.0   
4  17/Jul/2020:12:23:58 +0100  172.26.0.1   55903.0  172.26.0.4      80.0   
5  17/Jul/2020:12:24:07 +0100  172.26.0.1   55910.0  172.26.0.4      80.0   
6  17/Jul/2020:12:24:15 +0100  172.26.0.1   55911.0  172.26.0.4      80.0   

  request_http_method                               request_http_request  \
1                 GET  /blog/index.php/2020/04/04/voluptatum-reprehen...   
2                 GET                               /blog/xmlrpc.php?rsd   
4                 GET  /blog/index.php/2020/04/04/nihil-tenetur-et-ad...   
5                 GET  /blog/index.php/2020/04/04/explicabo-qui-fuga-...   
6                 GET  /blog/index.php/2020/04/04/explicabo-qui-fuga-...   

  request_http_protocol                                 request_user_agent  \
1              HTTP/1.1  Mozilla/5.0 (Macintosh; Intel Mac OS X 10.6; r...   
2              HTTP/1.1  Mozilla/4.0 (compatible; MSIE 6.0; Windows NT ...   
4              HTTP/1.1  Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_2...   
5              HTTP/1.1  Mozilla/5.0 (Windows NT 10.0; Win64; x64) Appl...   
6              HTTP/1.1  Mozilla/5.0 (Windows NT 5.1) AppleWebKit/537.3...   

  request_referer   request_host request_origin request_cookie  \
1             NaN  test-site.com            NaN            NaN   
2             NaN  test-site.com            NaN            NaN   
4             NaN  test-site.com            NaN            NaN   
5             NaN  test-site.com            NaN            NaN   
6             NaN  test-site.com            NaN            NaN   

  request_content_type request_accept request_accept_language  \
1                  NaN            */*                     NaN   
2                  NaN            */*                     NaN   
4                  NaN            */*                     NaN   
5                  NaN            */*                     NaN   
6                  NaN            */*                     NaN   

  request_accept_encoding  request_do_not_track request_connection  \
1           gzip, deflate                   NaN         keep-alive   
2           gzip, deflate                   NaN         keep-alive   
4           gzip, deflate                   NaN         keep-alive   
5           gzip, deflate                   NaN         keep-alive   
6           gzip, deflate                   NaN         keep-alive   

  request_body response_http_protocol  response_http_status_code  \
1          NaN               HTTP/1.1                      200.0   
2          NaN               HTTP/1.1                      200.0   
4          NaN               HTTP/1.1                      200.0   
5          NaN               HTTP/1.1                      200.0   
6          NaN               HTTP/1.1                      200.0   

  response_http_status_message  response_content_length  000 - Normal  \
1                           OK                  20862.0             1   
2                           OK                    317.0             1   
4                           OK                  20863.0             1   
5                           OK                  21062.0             1   
6                           OK                  21061.0             1   

   272 - Protocol Manipulation  242 - Code Injection  \
1                            0                     0   
2                            0                     0   
4                            0                     0   
5                            0                     0   
6                            0                     0   

   88 - OS Command Injection  126 - Path Traversal  66 - SQL Injection  \
1                          0                     0                   0   
2                          0                     0                   0   
4              

In [258]:
df_training = df_capec[(df_capec['label'] != '33 - HTTP Request Smuggling') & (df_capec['label'] != '248 - Command Injection') & (df_capec['label'] != '16 - Dictionary-based Password Attack') & (df_capec['label'] != '274 - HTTP Verb Tampering') & (df_capec['label'] != '88 - OS Command Injection')]

In [259]:
df_capec.shape

(599755, 41)

In [260]:
df_training.shape

(591798, 41)

In [263]:
# List of label
df_training.label.unique()

array(['000 - Normal', '126 - Path Traversal', '66 - SQL Injection',
       '272 - Protocol Manipulation',
       '310 - Scanning for Vulnerable Software', '242 - Code Injection',
       '153 - Input Data Manipulation', '194 - Fake the Source of Data',
       '34 - HTTP Response Splitting'], dtype=object)

In [265]:
df_final = df_training.loc[:, ['text','label']].reset_index(drop=True)

In [266]:
df_final.head()

text         label
0  GET /blog/index.php/2020/04/04/voluptatum-repr...  000 - Normal
1                           GET /blog/xmlrpc.php?rsd  000 - Normal
2  GET /blog/index.php/2020/04/04/nihil-tenetur-e...  000 - Normal
3  GET /blog/index.php/2020/04/04/explicabo-qui-f...  000 - Normal
4  GET /blog/index.php/2020/04/04/explicabo-qui-f...  000 - Normal

In [268]:
# Save to csv
df_final.to_csv('dataset_capec.csv', encoding='utf-8', index=False)